### ISSUE TITLE
Separation of multiple Optional Memorials falling on one day

### ISSUE THREAD

**Issue by tukusejssirs** [romcal/romcal#12]
In Slovakian calendar (see [this](https://lc.kbs.sk/?den=20180120)), on 20 Jan 2018 we could celebrate either and only one from the following:

1. Feria
2. Optional Memorial of St Fabian
3. Optional Memorial of St Sebastian
4. Votive Holy Mass in Honour to the Blessed Virgin Mary (literally translated as _Optional Memorial of Virgin Mary on Saturday_, for we in Slovakia have on every ferial Saturday this Optional Memorial).

This problem is with the string on the [78](https://github.com/pejulian/romcal/blob/cbf46aff82c85c36ddd848ce70bcbb18466bb137/src/locales/en.mjs#L78) line of `src/locales/en.mjs`, because it combine the two Optional Memorial of Sts Fabian and Sebastian into one.

Please consider to separate them for Slovakian calendar.

**Comment by tukusejssirs** [comment]
Another separation request: 3rd Feb: Saint Blase, Bishop Martyr/Saint Ansgar, Bishop:

1. Feria
2. Optional Memorial of St Blase
3. Optional Memorial of St Ansgar

src: https://lc.kbs.sk/?den=20170203

**Comment by tukusejssirs** [comment]
And yet another: 8th Feb: Saint Jerome Emiliani/Saint Josephine Bakhita, Virgin

1. Feria
2. Optional Memorial of St Jerome Emiliani
3. Optional Memorial of St Josephine Bakhita

src: https://lc.kbs.sk/?den=20170208

**Comment by pejulian** [comment]
Hi @tukusejssirs 

Thank you once again for the feedback,

The points raised here bring up the need for a new feature in `romcal` to properly handle the separation of feasts in `national` calendars. I given thought to what is needed to achieve this and plan to introduce a new feature called "composite dates" in the next release. I will include a new section in `README.md` to describe how composite dates work and how it resolves the issues that you have described above. 

To test the feature, I will make the feast separations as requested for `src/calendars/slovakia.mjs` along relevant with TDD tests,

God bless!

**Comment by tukusejssirs** [comment]
Okay, then. However, I thought the output would have two separate lines: one with St Fabian and the other with St Sebastian—and this would require two _separate_ keys. Probably a new option (argument) would be necessary to create an output as a 'combined row with all Optional Memorials' and another option with output of 'separate rows for all Optional Memorials.' Still, I think the latter would more readable and elegant, when there are 3 or more types of Masses possible.

Now it came to my mind, do you handle the ferial days, too? I mean, when there is no feast on a particular day OR there e only Optional Memorials, there should a Feria.

**Comment by tukusejssirs** [comment]
Actually, I think that `saintFabianPopeAndMartyrAndSaintSebastianMartyr` should be `saintFabianPopeMartyrAndMartyrAndSaintSebastianMartyr`, as St Fabian was Pope _and_ Martyr, too.

**Comment by pejulian** [comment]
@tukusejssirs 

If the given day has no feast or memorial and it is during ordinary time, the day will be defaulted to a Feria. This is achieved in the `src/lib/Seasons.mjs` file under the `earlyOrdinaryTime` or `laterOrdinaryTime` methods. The output will be something like `saturdayOfThe4thWeekOfOrdinaryTime`. This value is the base value that will be overridden by feasts or memorials from the general and national calendars.

**Comment by tukusejssirs** [comment]
In Slovakia, on 23rd Apr we celebrate [St Adalbert as Memoria](https://lc.kbs.sk/?den=20180423), therefore the celebration of St George had to be moved to [22nd Apr](https://lc.kbs.sk/?den=20180424), when we can celebrate:

(1) Feria
(2) St George
(3) St Fidelis

**Comment by tukusejssirs** [comment]
Just a side note: In addition to defaulting a day without feast/memorial to Feria it should do so also on optional memorial, i.e. there should be an output for feria **and** optional memorials at the same time.

**Comment by tukusejssirs** [comment]
Note to the separation: Saint Andrew Kim Taegon, Priest, and Paul Chong Hasang and Companions, Martyrs are celebrated (at least in Slovakia, but probably generally) together.

**Comment by tukusejssirs** [comment]
Note: I think this issue is a problem not only in Slovakian Liturgical Calendar, but in any calendar. Therefore I suggest to move it from project [Slovakian Liturgical Calendar](https://github.com/pejulian/romcal/projects/1) to [General fixes](https://github.com/pejulian/romcal/projects/2).

**Comment by tukusejssirs** [comment]
Another one:

$ grep -in Bede enUs.mjs 
129:    "saintBedeTheVenerablePriestAndDoctorSaintGregoryViiPopeSaintMaryMagdaleneDePazziVirgin": "Saint Bede the Venerable, Priest and Doctor/Saint Gregory VII, Pope/Saint Mary Magdalene de Pazzi, Virgin",
371:    "saintBedeTheVenerablePriestAndDoctor": "Saint Bede the Venerable, Priest and Doctor of the Church",

**Comment by tukusejssirs** [comment]
$ grep -in saintGertrudeTheGreatVirgin enUs.mjs 
232:    "saintMargaretOfScotlandSaintGertrudeTheGreatVirgin": "Saint Margaret of Scotland/Saint Gertrude the Great, Virgin",
458:    "saintGertrudeTheGreatVirgin": "Saint Gertrude The Great, Virgin",

**Comment by tukusejssirs** [comment]
$ grep -in saintAidanBishop enUs.mjs 
382:    "saintAidanBishopAndTheSaintsOfLindisfarne": "Saint Aidan, Bishop and The Saints of Lindisfarne",
507:    "saintAidanBishop": "Saint Aidan, Bishop",

**Comment by tukusejssirs** [comment]
@pejulian, what are your thoughts on this? Your decision is in my opinion fundamental, because if we are to implement issue #20 (lit-cal-style output), we need to have some portable way to accomplish this. And also, in sorting keys (issue #18), I’d like to create a separate group with keys to be separated if it is going  to be implemented.

You need to decide if (1) we are to separate all celebrations (even those of which we do not know yet if we need them to be separated) and, then, those needed to be combined would be combined in-place; or (2) we are to separate only those celebrations of which we already know that some national cals use them separately; (3) something else.

Keep in mind that if we are to implement the lit-cal-style generation (issue #), my first suggestion would work better (even though it would require a bit more to re-code). And also, it would need to handle multiple optional memorials, as well as ferial on the same day.

**Comment by tukusejssirs** [comment]
$ grep -inA2 saintMargaretOfScotland ../calendars/*
../calendars/england.mjs:393:      "key": "saintEdmundOfAbingdonBishopOrSaintMargaretOfScotland",
../calendars/england.mjs-394-      "type": Types[6],
../calendars/england.mjs-395-      "moment": moment.utc({ year: year, month: 10, day: 16 }),
--
../calendars/general.mjs:1584:      "key": "saintMargaretOfScotlandSaintGertrudeTheGreatVirgin",
../calendars/general.mjs-1585-      "type": Types[6],
../calendars/general.mjs-1586-      "moment": moment.utc({ year: year, month: 10, day: 16 }),
--
../calendars/scotland.mjs:122:      "key": "saintMargaretOfScotland",
../calendars/scotland.mjs-123-      "type": Types[4],
../calendars/scotland.mjs-124-      "moment": moment.utc({ year: year, month: 10, day: 16 }),

**Comment by tukusejssirs** [comment]
```bash
$ grep -n "Canute\|Eric\|Olaf" enUs.mjs 
400:    "saintEricIxMartyr": "Saint Eric IX, Martyr",
405:    "saintCanuteMartyr": "Saint Canute, Martyr",
407:    "saintOlafIiMartyr": "Saint Olaf II, Martyr",
440:    "saintsCanuteEricAndOlafMartyrs": "Saints Canute, Eric and Olaf, Martyrs",  # Joint celebration in Germany
```

**Comment by tukusejssirs** [comment]
Another way to describe the joint optional memorials issue, is using a generated [German Calendar](http://www.eucharistiefeier.de/lk/api/lko0000.html). In there, search for the 3rd Nov 2018 (`03.11.2018`).

There, you can see the following optional memorials and ferial (translated into English):
```
Hubert of Liège, Bishop
Martin de Porres, Religious
Pirmin, Abbot, Bishop and Missionary on the Upper Rhine
Rupert Mayer, Priest
Saturday of the 30th week of Ordinary Time
```

There are 5 different celebrations that might occur on one particular day.

**Comment by tukusejssirs** [comment]
This a multi-step / complex thing. The following algorithm (and step-by-step todo list) is my proposal, you don’t need to agree with it nor follow it.

## Create `join` 'method'
I think this is the easiest to do (for you). You would to create a 'method' (or whatever you call it) like the `drop` one. `join` would be used to 'smart-merge' selected saints/celebration when needed.

## Create a list of saints/celebrations
The process of separation is so complex because we to process a list of saints/celebrations, which need need to create first. This list should in `src/locales/`,

This list contain contain (1) saint/celebration identifier (`vianneyJohnMary`) and its properties; for example:

```
vianneyJohnMary{
    .surname="Vianney"  // It might be a [nobiliary particle](https://en.wikipedia.org/wiki/Nobiliary_particle), too
    .name="John Mary"  // All first names
    .holiness="Saint"  // This might be an int; todo: find better name instead of 'holiness'; options: st/sts/bl/bls
    .title="priest"  // Comma-separated list
    .number=1  // How many persons?; only used in `saints/`
    .sex="male"  // male/female/mix; only used in `saints/`
}
```

This first two (`id.{surname,name}`) would be used directly to generate a name of celebration, while the third and fourth one (`id.{holiness,title}`) depends `id.{number,sex}` and only then they would be incorporated into the name. This list of saints would substitute the current `src/locales/*mjs` files together with the list of titles (see below).

## Create a list of titles
We also need a list of titles (martyr, priest, abott, etc) in singular masculine, singular feminine, plural masculine, plural feminine. Some of them might be removed / left empty (like feminine forms a priest/bishop/pope). The reason for two plural forms is that in some languages (I know of Hebrew, at least the Biblical/Ancient one), they have two plural forms (masculine and female); however, even in English we sometimes need all four forms (see `abbot` example).

Also this list should be in `src/locales/`.

Example:

```
priest{
    .sgm="priest" // SinGular Masculine
    .sgf="" // Might be omitted
    .plm="priests"
    .plf="" // Might be omitted
}

abbot{
    .sgm="abbot"
    .sgf="abbess"
    .plm="abbots"
    .plf="abbesses"
}
```

In the title generation, it would be nice to have check if one has been to priests and popes—in that case, priest would be omitted in favour of a higher-ranked title. 'Ranks' order: pope > bishop > priest > deacon. Hopefully, this would be never used.

Also, the title id (`priest` or `abbot`) should be in a form of `{id}.sgm`, but for the sake of sanity, we should be able to detect (read: 'deciphre') any of the four corms in saint title property. One way to achieve it, would be to auto-create an auxiliary file that would contain all other forms (sgf, plm, plf) as id’s with all the same title properties, but that would be a quite heavy workaround. Maybe you could make up a better way.

## Celebrations in calendars
There would be a change in `src/calendars/*.mjs`, too. If nothing needs to be dropped, joined and a national key (celebration name) is not needed, it would be quite simple (that is: the way it is)—you would just use the key id `vianneyJohnMary` where you need it. The same would apply to dropping a celebration.

If one needs to join two saints into one celebration, s/he would need to write soththing like this (this is just a pseudo-code, the actual code might be different):
```
join{
    .celebrations="peterApostle,paulApostle" // Comma-separated list
    .additionalTitles="" // This property is optional; it might be used when there’s an additional title added the joined celebration name in contrast to all the ‘separated’ celebrations; this might also be used for adding ‘patron of {continent}’ to the celebration name, but then its name (`join`) should be changed to something else
    .isHolyDay=0 // Whether it is [holy day of observation](https://en.wikipedia.org/wiki/Holy_day_of_obligation) or not
    .nameOverride=""  // Used only when the name of joined celebration is totally different from the name of the saints (e.g. celebration of `saintsMarkoKrizinMelicharGrodeckiAndStephenPongracPriestsAndMartyrs` in Slovakia we celebrate as `Three saints Martyrs of Košice—Marko Krizin, Melichar Grodecki and Stephen Pongrac, Priests and Martyrs`); we should be able to use some variables in this (like `Three saints Martyrs of Košice—$name1, $name2, $name3, $title`; this should generate the same text as I’ve written above)
}
```

**Comment by tukusejssirs** [comment]
So, I started to seperate saints (at least extra repo) in a [Google Sheet](https://docs.google.com/spreadsheets/d/1c9YWtfntpxy0r0rg0mfowJ2v-yKW6wSMmNzyY3umBU4/edit?usp=sharing). I have not finished it and there are still some stuff to deal with.

I post it here in order to let anyone and everyone know that I’d still like this to implement into `romcal`.

This might be useful for Tridentine Calendar, too.

@pejulian @bblomf @agbloomfield

**Comment by emagnier** [comment]
@tukusejssirs, I'm very impressed by all your big work in the Google Sheet document! 
I also like the idea to have separated keys per saints, and be able to output multiple entities for the same day (as well as the Feria, particularly when we have multiple optional memories).
Also because sometime things move or a new saint appears in the calendar.

All your thoughts are continuing to convince me that we should move to a DB, and think about what fields/properties we will need for each saint/celebration...

**Comment by tukusejssirs** [comment]
> @tukusejssirs, I'm very impressed by all your big work in the Google Sheet document!

Well, everyone does what they can do best! Anyway, thanks! Still that work is not done yet (it needs to be finished and then ported/implemented to `romcal`).

> I also like the idea to have separated keys per saints, and be able to output multiple entities for the same day (as well as the Feria, particularly when we have multiple optional memories).

Yes, the ferias were the first issue I wanted to deal with, but later on (during the _separation_ process) I have found other issues, some of which I have reported in the GitHub issues, some are only dealt in the Gootle Sheets.

> All your thoughts are continuing to convince me that we should move to a DB, and think about what fields/properties we will need for each saint/celebration...

Agreed. We should make the port to DB in the next major version of `romcal` (`v2.0.0`), as this breaks (let’s say) the current API.

Some notes to the DB fields/properties (note that I have no idea what is the proper way to use DBs for translations, but I use DBs, like PostgreSQL):
- I would suggest to keep in mind all the translation _sections_ (`saints`, `celebrations`, `titles`, `canonisation`, `companions`, `seasons`, `ordinary`, `numbers`, `ui`; I would even add `calendars` to list/translate the names of the calendars, see #23; and also `other` to translate other strings like some strings from the _directory_, see #20) should be either in separate database or separate schema (schema might be enough). This would make it well arranged / transparent.
- Currently, we use `lowerCamelCase` as a celebration ID. When porting to DB, I might be better to use a unique integer (ID/serial) to identify a saint/celebration. Of couse, this ID would be dependent on a DB table, which is (in my opinion) okay. And this would be true for all _sections_.

**Comment by emagnier** [comment]
Trying to summarise... On a specific date:

- All elements should be outputted as soon as we don't have something higher than `OPT_MEMORIAL`. 
- If an element is a `MEMORIAL`, other type ranks under `MEMORIAL` are dropped (`OPT_MEMORIAL`, `COMMEMORATION`, `WEEKDAY`). This is actually how it's implemented.
- Saints that should be celebrated together should stay one "item", because an item represent one celebration.

But... a few comments:

- When we have only `OPT_MEMORIAL` or `COMMEMORATION` items and the `WEEKDAY` item on a specific date, `WEEKDAY` should be returned first, since it will be the "default" celebration as long as we don't decide to celebrate an optional memorial (note: `romcal` always output items in an array, even if there is only one item).

- The `COMMEMORATION` types are basically all `FEAST`, `MEMORIAL` and `OPT_MEMORIAL` deranked to commemoration when they are occuring in the season of LENT. This is actually well implemented in romcal.

- When `WEEKDAY` isn't outputted, it might be still useful to have the feria information somewhere. Particularilly for the liturgy of the hours, and because if something is missing in a celebration (hymn, antiphon, verses / responses, prayers...), it's recommanded to take the ones from the commons. The feria information could be added in a `data.feria` property for exemple. Same thing for a `SOLEMNITY` taking over a `SUNDAY` (Sunday data could be added in `data.sunday`).

**Comment by tukusejssirs** [comment]
> Saints that should be celebrated together should stay one "item", because an item represent one celebration.
Sort of yes and no. At least not in the first place. Let me try to be more specific.

1. I suggest that for _localisation_ purposes, we should separate all saints joint celebrations into separate saint.
2. Then, in the calendar settings (i.e. a file like `france.js`), the ‘calendarists’ would join the saints using a `join()` function (or whatever thing you choose) into a joint celebration.
3. Finally (when all previous steps are completed), yes, _joint celebrations of saints should ~stay~ be output as one celebration_.

> When WEEKDAY isn't outputted, it might be still useful to have the feria information somewhere.

I have no idea if there is similar book to directory / liturgical calendar/guide in French (for more info, see #20), but in the Slovak Directories (published each year for period from the First Advent Sunday to the end of the following civil year), we have notes on _celebrations not celebration in that particular year_. This happens when a higher-ranked celebration (e.g. a Sunday) occurs on a day when there is a memoria. We _might_ include this for ~WEEKDAY~ FERIA, but I think this is not necessary. It is enough to have a (sort of) API to get that info (as you suggest) which we could use to fix issue #20. If you still want to output this kind of note, make it optional (i.e. not default) and create a new option for that (as kind of a toggle).

**Comment by emagnier** [comment]
Concerning the first part of your comment, combine several text elements will be something possible when a i18n library will be integrated in romcal (with the project of improving locale management, multiple plural support, etc.).
In the calendar part, this need to stay one celebration (containing the combinaison of text elements).
I guess it's what you mean.

And concerning the second part, I also agree with you. I just want to highlight that needs/rules for a mass and the liturgy of the Hours could be different. Adding the SUNDAY/WEEKDAY information on theses specific celebrations (as an object property on the celebration item) seems reasonable to address theses cases. It will change anything for someone just interested by getting calendars for mass.

>~WEEKDAY~ FERIA

Do you mean here we should rename WEEKDAY to FERIA? It tempts me too :)

Also WEEKDAY is currently the last type item, just behind COMMEMORATION.
I'm wondering if COMMEMORATION should be the last item instead of WEEKDAY.

**Comment by tukusejssirs** [comment]
> Concerning the first part of your comment, combine several text elements will be something possible when a i18n library will be integrated in romcal (with the project of improving locale management, multiple plural support, etc.).
> In the calendar part, this need to stay one celebration (containing the combinaison of text elements).
> I guess it's what you mean.

I agree this will be possible only _after_ the i18n will be integrated into `romcal`. But I disagree on the calendar part. Currently, we have multiple celebrations of the very same saints (a joint celebration in one calendar, separate celebrations in others). This issue won’t be fully solved until there is only one translation of a particular saint. For example, read [this comment of mine](https://github.com/romcal/romcal/issues/12#issuecomment-380150421).

> Do you mean here we should rename WEEKDAY to FERIA? It tempts me too :)

Yes. But I am not only tempted to do so, but I even opened a new issue #88 (to make to note to myself at least). A somewhat related issue is #73 to change the _national_ calendars to _particular_ calendars, as in the (English) liturgical books, these terms are used.

> Also WEEKDAY is currently the last type item, just behind COMMEMORATION.
> I'm wondering if COMMEMORATION should be the last item instead of WEEKDAY.

**Comment by tukusejssirs** [comment]
> In the calendar part, this need to stay one celebration (containing the combinaison of text elements).
I guess it's what you mean.

I sort of did not read these your words. Yeah, we might think of the same thing from different perspectives. :smiley:

Indeed, also I want that the calendars would contain (where needed) a _single joint celebration of multiple saints_, but the definition should also be in the calendar itself. For example, when a joint celebration is (let’s say_ introduced in the `general.js`, but in a particular calendar those saints (celebration jointly in `general.js`) are celebrated separately, these can be simply redefined by:
- dropping the celebration defined in `general.js`;
- defining separate celebrations for the saints.

**Comment by emagnier** [comment]
Yes we could have a sort of "join" option to combine together the name of different saints from locale files. This might be something not too difficult with the functions offered by a i18n library.
But I'm not in favor to join 2 celebrations from calendars (not locales) together. Saints that are celebrated together still need to be a unique celebration (with a unique key), with its unique celebration properties.

**Comment by tukusejssirs** [comment]
> But I'm not in favor to join 2 celebrations from calendars (not locales) together.

What do you mean by _celebrations from calendars_?

All I wanted to say, that the saints should be ‘created’ somewhere (like `src/persons.js` I have suggested elsewhere) and in the calendars, these saints should be joined into joint celebrations as needed.

For example: _St Paul, apostle_ and _St Peter, apostle_ should be (separately) defined in the `persons.js` file. Then, in the `general.js` calendar, the _Sts Peter and Paul, apostles_ celebration should be defined by joining the two saints.

> Saints that are celebrated together still need to be a unique celebration (with a unique key), with its unique celebration properties.

I somewhat disagree. Why would a joint celebration need a unique key or unique celebration properties? I don’t get it.

Still, if you think that we need it, we might create these properties while joining the saints into that particular celebration. :smiley:

**Comment by jarosz** [comment]
Milestone is 1.3.0, but there's a lot of ideas for future development here.
Is combining multiple optional memorials into one separate key with "Or", like "saintCamillusDeLellisPriestOrSaintHenryBishopAndMartyr" still relevant for 1.3.0?

**Comment by emagnier** [comment]
Right that one is more a discussion with many ideas. I added it by mistake in 1.3.0.
We can consider #74 to be relevant for 1.3.0

**Comment by tukusejssirs** [comment]
I don’t if you all agree with me, but I think this _separation_ should be dealt with the implementaion of i18n, and therefore I suggest to move this issue to `v2.0.0` milestone.

**Comment by emagnier** [comment]
This has been implemented in #162.
